### ISSUE TITLE
O_EMPTYPATH

### ISSUE THREAD

**Issue by cyphar** [openSUSE/libpathrs#16]
This is required to complete #7 -- because it's possible to bind-mount over magic-links we cannot trust `/proc/self/fd/...` and thus must use a proc-less method to re-open file descriptors.

**Comment by LATENTBOUNCE** [comment]
O_EMPTYPATH is upstream since 7.2, so I guess this can be closed? Never thought about the bind-mount attack surface here... I have been looking into libpathrs a bit and how I could contribute, because I'm also interested in path resolution ever since trying to get O_CREAT|O_DIRECTORY in.

Best,
Jori (Sorry for the ridiculous name, I don't know how to change it to just my real name in Github...)

**Comment by cyphar** [comment]
Well, once libpathrs is migrated to using it (with back-compat shims, of course) we can close it. ;)

The bind mounts over magic-links problem was mostly solved in 6.12 with torvalds/linux@d80b065bb172 but this is still a problem on older kernels and is a bit of a tricky problem with the handful of other magic-links that were not touched by that change.

(You should be able to rename your account in your profile or somewhere in user settings?)

**Comment by LATENTBOUNCE** [comment]
Yes, OK, that sounds good. Do you have an example of how such a back-compat shim should look like? I can make an educated guess, but I would rather not run into any pitfalls. And when/how does the shim get dropped? I really like this area of interaction between kernel UAPI and low level userspace. I know, for instance, systemd also has chaseat() which (I think) does not use all new kernel features yet.

Yeah, I already changed my username, but it still shows up as this randomized username I once chose here.

PS. Will you be at Plumbers this year?

**Comment by cyphar** [comment]
> Yes, OK, that sounds good. Do you have an example of how such a back-compat shim should look like? I can make an educated guess, but I would rather not run into any pitfalls.

`ProcfsHandleBuilder::build` is the most "complete" example but the general idea is that you should try to use the new functionality and fallback to the old one if it doesn't work, this then needs to be paired with tests that exercise it (ideally directly -- we have `fake-enosys` but that won't work for these flags with `openat2` because you can't seccomp-filter pointers yet -- though maybe we could use `openat` for now?).

> And when/how does the shim get dropped?

Once we bump our minimum kernel version but I don't expect that to happen soon, and especially not to 7.2. On paper, we support kernels back to 2.6.39 (though we don't test that far back at the moment).

> I know, for instance, systemd also has chaseat() which (I think) does not use all new kernel features yet.

They try to use newer stuff but it needs to be able to have a graceful fallback on whatever their minimum kernel version is -- IIRC it's 5.10 or something now? That's basically the same philosophy as libpathrs (and most projects that need to deal with low-level kernel APIs).

> PS. Will you be at Plumbers this year?

Yup!
